### ISSUE TITLE
bug in your VAE cost

### ISSUE THREAD

**Issue by jmaronas** [wiseodd/generative-models#43]
loss = recon_loss + kl_loss

While you should have
loss = recon_loss - kl_loss

To perform gradient descent.

If not you should
loss = -recon_loss + kl_loss
and perform gradient ascent.

The ELbo to maximize is:
ELBO = -DKL + LogLikelihood

Your kl_loss correspond to the -DKL and your binary error to -LogLikelihood. So you have to change your -DKL to be DKL to minimize the ELBO, if not change sign from LogLikelihood and maximize.

**Comment by wiseodd** [comment]
Hi, 

`kl_loss` is D_KL, not -D_KL.

Please compare:
``` python
recon_loss = nn.binary_cross_entropy(X_sample, X, size_average=False) / mb_size
kl_loss = torch.mean(0.5 * torch.sum(torch.exp(z_var) + z_mu**2 - 1. - z_var, 1))
loss = recon_loss + kl_loss
```

found in this repo's code, with:

```python
def vae_loss(x, x_decoded_mean):
    xent_loss = objectives.binary_crossentropy(x, x_decoded_mean)
    kl_loss = - 0.5 * K.mean(1 + z_log_sigma - K.square(z_mean) - K.exp(z_log_sigma), axis=-1)
    return xent_loss + kl_loss
```

As found in https://blog.keras.io/building-autoencoders-in-keras.html.

**Comment by jmaronas** [comment]
well you can actually trust other people implementation or  read the paper:
The Elbo cost is defined as:

ELBO = -DKL + log(p(x|z)

If you want to minimize the cost, that is, perform gradient descent you have to:

-ELBO = +DKL - log(p(x|z)

For this purpose, nn.binary_cross_entropy returns the -log(p(x|z)), however, as stated in the paper:
kl_loss = torch.mean(0.5 * torch.sum(torch.exp(z_var) + z_mu**2 - 1. - z_var, 1))
is the -DKL (see appendix B on the paper of VAEs)

So this mean you have to either multipy by -1 your nn.binary_cross_entropy and perform gradient ascent (just change sign in your optimizer) or multiply by -1 your kl_loss and perform gradient descend.

You can confirm what I tell you looking at the pytorch example.
https://github.com/pytorch/examples/blob/master/vae/main.py

I think the keras implementation is not good either.

**Comment by jmaronas** [comment]
Actually I have implement the VAE, you can see the generative process. Just change the sign. I look several implementation on VAEs that are wrong also. However I like your repo it have very interesting things so If you could fix that would be good.

I am trying to look if keras objectives.binary_crossentropy returns negative or positive. However I am not keras user so I am not used to it. Probably if that is correctly implemented should return the positive value, if not, that would be also wrong. Look at this

https://wiseodd.github.io/techblog/2016/12/10/variational-autoencoder/

This  is in keras an it is different. I think this is the correct one for the fact that probably the binary_crossentropy from keras is negative value.
![sample_19](https://user-images.githubusercontent.com/15247250/33438289-ad778a10-d5ea-11e7-8b67-6b6e375dccac.png)
![sample_0](https://user-images.githubusercontent.com/15247250/33438293-af81350e-d5ea-11e7-96e0-8ee2dd18c6b8.png)

**Comment by wiseodd** [comment]
```python
kl_loss = torch.mean(0.5 * torch.sum(torch.exp(z_var) + z_mu**2 - 1. - z_var, 1))
```
is actually D_KL, not -D_KL. Try to compare it with result in Appendix B: they have the opposite sign.

**Comment by jmaronas** [comment]
you are totally right. I get confused because some people put sign inside others do not.
